### DISCUSSION TITLE
Unit Testing

### DISCUSSION THREAD

**Discussion by elcarn23** [cockpit-project/cockpit#16892] [Q&A] [answered]
Is it possible to use unit testing when developing a plugin for cockpit?  When running tests the code which works while viewing the plugin in cockpit can no longer import cockpit to be used with the unit test.

**Comment by martinpitt** [comment]
Yes, it is possible, and most practical for testing logic. Cockpit has QUnit JS unit tests for the [Storage](https://github.com/cockpit-project/cockpit/blob/main/pkg/storaged/test-util.js) and [Networking](https://github.com/cockpit-project/cockpit/blob/main/pkg/networkmanager/test-utils.js) pages for example, and tons of unit tests for [cockpit.js itself](https://github.com/cockpit-project/cockpit/tree/main/pkg/base1).

This requires some plumbing: The [qunit-* files in pkg/lib](https://github.com/cockpit-project/cockpit/tree/main/pkg/lib), [webpacking them](https://github.com/cockpit-project/cockpit/blob/a8e17efc7af6a2dd4e5c782f96bfe8b607713055/webpack.config.js#L366), and [some plumbing in the Makefile](https://github.com/cockpit-project/cockpit/blob/a8e17efc7af6a2dd4e5c782f96bfe8b607713055/Makefile.am#L125). It also needs a http server -- qunit has some provisioning for this, but for cockpit we use our own (as it ships a web server anyway).

So it is certainly *possible* but a lot of effort to set up. That's why starter-kit does not ship any unit tests, as in most cases for cockpit plugins, it's not worth the effort -- integration tests are more useful in general, and even easier to write with our test machinery.

**Comment by martinpitt** [comment, marked as answer]
Actually, have a look at [cockpit-ostree](https://github.com/cockpit-project/cockpit-ostree/tree/main) -- that also does qunit unit tests, but in a much simpler way. [The test](https://github.com/cockpit-project/cockpit-ostree/blob/main/src/test-utils.js) does not have anything weird, the webpack.config.js builds it in a straightforward way, and package.json has script starters for it.

**Reply by elcarn23** [reply]
Thank you for the response.  I will take a look at the information provided.
